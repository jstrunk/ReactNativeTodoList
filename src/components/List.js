import React, {Component} from 'react';
import {FlatList,Text,StyleSheet,TouchableOpacity} from 'react-native';
import { ListItem, CheckBox } from 'react-native-elements';

const styles = StyleSheet.create({
  list: {
  },
  item: {
    textAlign: 'left',
    color: 'black',
  },
  remove: {
    textAlign: 'center',
    color: 'red',
  },
});

export default class List extends Component {


  renderItem = ({item, index}) => (
    <ListItem
      title={item.key}
      titleStyle={styles.item}
      checkBox={{
        right: true,
        checked: this.props.items[index].done,
        onIconPress: () => {
          let items = [...this.props.items];
          let item = {...items[index]};
          item.done = !item.done;
          items[index] = item;
          this.props.onUpdate(items);
        },
      }}
      rightElement={
        <TouchableOpacity onPress={this.removeItem(index)}>
          <Text style={styles.remove}>X</Text>
        </TouchableOpacity>
      }
    />
  );

  removeItem(index) {
    return () => {
      let items = [...this.props.items];
      items.splice(index,1);
      this.props.onUpdate(items);
    }
  }

  render() {
    return (
      <FlatList
        style={styles.list}
        data={this.props.items}
        renderItem={this.renderItem}
      />
    );
  }
}