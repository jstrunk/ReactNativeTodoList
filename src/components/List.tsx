import React, {Component} from 'react';
import {FlatList,StyleSheet} from 'react-native';
import { ITodoList } from '../redux/types';
import TodoItem from '../containers/TodoItem';

interface Props {
  items: ITodoList;
}

interface ItemProps {
  item: string;
  index?: number;
  separators?: any;
}

const styles = StyleSheet.create({
  list: {
    marginBottom: 40,
  },
  item: {
    textAlign: 'left',
    color: 'black',
  },
  remove: {
    textAlign: 'center',
    color: 'red',
  },
  checkBox: {
    color: '#bfbfbf',
  },
});

export default class List extends Component<Props> {
  renderItem = (itemProps: ItemProps): any => (
    <TodoItem id={itemProps.item} styles={styles}/>
  );

  render(): JSX.Element {
    return (
      <FlatList
        style={styles.list}
        data={this.props.items}
        renderItem={this.renderItem}
        keyExtractor={(item, index) => index.toString()}
      />
    );
  }
}
