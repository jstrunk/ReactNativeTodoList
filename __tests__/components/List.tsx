import 'react-native';
import React from 'react';
import List from '../../src/components/List';

import { render, fireEvent } from '@testing-library/react-native'

describe('List component', () => {
  const items = [
    {
      key: 'Run the tests',
      done: false,
      id: 'abc1',
    },
    {
      key: 'foo',
      done: true,
      id: 'def2',
    },
  ];

  const mockToggleItem = jest.fn();
  const mockDeleteItem = jest.fn();


  test('renders correctly', () => {
    const { asJSON } = render(
      <List items={items} toggleItem={mockToggleItem} deleteItem={mockDeleteItem} />);
    expect(asJSON()).toMatchSnapshot();
  });

  it('has items', () => {
    const { getAllByText } = render(
      <List items={items} toggleItem={mockToggleItem} deleteItem={mockDeleteItem} />
    );

    const fooElements = getAllByText('foo');
    const checkedElements = getAllByText('');
    const deleteElements = getAllByText('X');
    expect(fooElements).toHaveLength(1);
    expect(checkedElements).toHaveLength(1);
    expect(deleteElements).toHaveLength(2);

    checkedElements.forEach((e) => fireEvent.press(e));
    deleteElements.forEach((e) => fireEvent.press(e.parentNode));

    expect(mockToggleItem).toHaveBeenCalledTimes(1);
    expect(mockDeleteItem).toHaveBeenCalledTimes(2);
  });



});
