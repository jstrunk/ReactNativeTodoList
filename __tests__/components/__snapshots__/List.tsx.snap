// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      activationDistance={0}
      animationConfig={
        Object {
          "damping": 20,
          "mass": 0.2,
          "overshootClamping": false,
          "restDisplacementThreshold": 0.2,
          "restSpeedThreshold": 0.2,
          "stiffness": 100,
        }
      }
      autoscrollSpeed={100}
      autoscrollThreshold={30}
      collapsable={false}
      data={
        Array [
          "abc1",
          "def2",
        ]
      }
      disableVirtualization={false}
      extraData={
        Object {
          "activeKey": null,
          "extraData": undefined,
          "hoverComponent": null,
        }
      }
      horizontal={false}
      initialNumToRender={10}
      maxToRenderPerBatch={10}
      onEndReachedThreshold={2}
      removeClippedSubviews={false}
      scrollEnabled={true}
      scrollEventThrottle={1}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "marginBottom": 40,
        }
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "translateY": AnimatedValue {
                    " __value": 0,
                  },
                },
              ],
            }
          }
        >
          <View
            pointerEvents="auto"
            style={
              Object {
                "flexDirection": "column",
              }
            }
          >
            <View>
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "white",
                      "borderColor": "#bcbbc1",
                      "flexDirection": "row",
                      "padding": 14,
                    }
                  }
                  testID="padView"
                >
                  <View
                    style={
                      Object {
                        "alignItems": "flex-start",
                        "flex": 1,
                        "justifyContent": "center",
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "backgroundColor": "transparent",
                          "color": "black",
                          "fontSize": 17,
                          "textAlign": "left",
                        }
                      }
                      testID="listItemTitle"
                      theme={
                        Object {
                          "colors": Object {
                            "disabled": "hsl(208, 8%, 90%)",
                            "divider": "#bcbbc1",
                            "error": "#ff190c",
                            "grey0": "#393e42",
                            "grey1": "#43484d",
                            "grey2": "#5e6977",
                            "grey3": "#86939e",
                            "grey4": "#bdc6cf",
                            "grey5": "#e1e8ee",
                            "greyOutline": "#bbb",
                            "platform": Object {
                              "android": Object {
                                "error": "#f44336",
                                "primary": "#2196f3",
                                "secondary": "#9C27B0",
                                "success": "#4caf50",
                                "warning": "#ffeb3b",
                              },
                              "ios": Object {
                                "error": "#ff3b30",
                                "primary": "#007aff",
                                "secondary": "#5856d6",
                                "success": "#4cd964",
                                "warning": "#ffcc00",
                              },
                            },
                            "primary": "#2089dc",
                            "searchBg": "#303337",
                            "secondary": "#8F0CE8",
                            "success": "#52c41a",
                            "warning": "#faad14",
                          },
                        }
                      }
                    >
                      abc1
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "translateY": AnimatedValue {
                    " __value": 0,
                  },
                },
              ],
            }
          }
        >
          <View
            pointerEvents="auto"
            style={
              Object {
                "flexDirection": "column",
              }
            }
          >
            <View>
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "white",
                      "borderColor": "#bcbbc1",
                      "flexDirection": "row",
                      "padding": 14,
                    }
                  }
                  testID="padView"
                >
                  <View
                    style={
                      Object {
                        "alignItems": "flex-start",
                        "flex": 1,
                        "justifyContent": "center",
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "backgroundColor": "transparent",
                          "color": "black",
                          "fontSize": 17,
                          "textAlign": "left",
                        }
                      }
                      testID="listItemTitle"
                      theme={
                        Object {
                          "colors": Object {
                            "disabled": "hsl(208, 8%, 90%)",
                            "divider": "#bcbbc1",
                            "error": "#ff190c",
                            "grey0": "#393e42",
                            "grey1": "#43484d",
                            "grey2": "#5e6977",
                            "grey3": "#86939e",
                            "grey4": "#bdc6cf",
                            "grey5": "#e1e8ee",
                            "greyOutline": "#bbb",
                            "platform": Object {
                              "android": Object {
                                "error": "#f44336",
                                "primary": "#2196f3",
                                "secondary": "#9C27B0",
                                "success": "#4caf50",
                                "warning": "#ffeb3b",
                              },
                              "ios": Object {
                                "error": "#ff3b30",
                                "primary": "#007aff",
                                "secondary": "#5856d6",
                                "success": "#4cd964",
                                "warning": "#ffcc00",
                              },
                            },
                            "primary": "#2089dc",
                            "searchBg": "#303337",
                            "secondary": "#8F0CE8",
                            "success": "#52c41a",
                            "warning": "#faad14",
                          },
                        }
                      }
                    >
                      def2
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
